fdbrpc/DDSketch.h

```diff
@@ -0,0 +1,190 @@
+	static inline uint64_t doubleToLongBits(double x) {
+		uint64_t u;
+		uint64_t* pu = reinterpret_cast<uint64_t*>(&x);
+		u = *pu;
+		return u;
+	}
+	static inline double longBitsToDouble(uint64_t x) {
+		double u;
+		double* pu = reinterpret_cast<double*>(&x);
+		u = *pu;
+		return u;
+	}
+	static int getExponent(uint64_t longBits) {
+		return (int)((longBits & EXPONENT_MASK) >> EXPONENT_SHIFT) - EXPONENT_BIAS;
+	}
+
+	static double getSignificandPlusOne(uint64_t longBits) {
+		return longBitsToDouble((longBits & SIGNIFICAND_MASK) | ONE);
+	}
+	static double fastlog(double value) {
+		uint64_t longBits = doubleToLongBits(value);
+		double s = getSignificandPlusOne(longBits) - 1;
+		double e = getExponent(longBits);
+		return ((A * s + B) * s + C) * s + e;
+	}
+};
+#endif
+
+// A DDSketch for non-negative numbers
+// (those < 10^-18 are treated as 0, and huge numbers (>>10^18) may degrade the performance)
+template <class T>
+class DDSketch {
+public:
+	explicit DDSketch(double errorGuarantee = 0.1)
+	  : errorGuarantee(errorGuarantee), gamma((1.0 + errorGuarantee) / (1.0 - errorGuarantee)), logGamma(log(gamma)),
+	    multiplier(FASTLOG ? fastLogger::correctingFactor * log(2) / log(1 + 2 * errorGuarantee / (1 - errorGuarantee))
+	                       : 1),
+	    offset((FASTLOG ? fastLogger::fastlog(1 / EPS) * multiplier : log(1 / EPS)) + 5), populationSize(0),
+	    zeroPopulationSize(0), _min(T()), _max(T()), sum(T()) {
+
+		buckets.resize(2 * offset, 0);
+#if FASTLOG
+		ASSERT(sizeof(double) == sizeof(uint64_t));
+#endif
+	}
+
+	DDSketch<T>& addSample(T sample) {
+		// Call it addSample for now, while it is not a sample anymore
+		if (!populationSize) _min = _max = sample;
+
+		if (sample <= EPS) {
+			zeroPopulationSize++;
+		} else {
+#if FASTLOG
+			int index = ceil(fastLogger::fastlog(sample) * multiplier);
+#else
+			int index = ceil(log(sample) / logGamma);
+#endif
+			ASSERT(index + offset >= 0);
+			if (index + offset >= buckets.size()) {
+				buckets.resize(index + offset + 1);
+			}
+			buckets[index + offset]++;
+		}
+
+		populationSize++;
+		sum += sample;
+		_max = std::max(_max, sample);
+		_min = std::min(_min, sample);
+		return *this;
+	}
+
+	double mean() const {
+		if (populationSize == 0) return 0;
+		return (double)sum / populationSize;
+	}
+
+	T median() { return percentile(0.5); }
+
+	T percentile(double percentile) { // Find the targetPercentilePopulation-th element
+		if (percentile < 0.0 || percentile > 1.0 || populationSize == 0) return T();
+		uint64_t count = zeroPopulationSize;
+		uint64_t targetPercentilePopulation = percentile * (populationSize - 1);
+		if (targetPercentilePopulation < zeroPopulationSize) {
+			// Zeros are enough
+			return (T)EPS;
+		}
+		for (size_t i = 0; i < buckets.size(); i++) {
+			if (targetPercentilePopulation < count + buckets[i]) {
+#if FASTLOG
+				return (T)(2.0 * pow(gamma, (i - offset) / fastLogger::correctingFactor) / (gamma + 1));
```

Comment:
If we don't go with using a fixed error size, this should probably use the `inverseLog` method used by `CubicallyInterpolatedMapping.java`: https://github.com/DataDog/sketches-java/blob/97c4a8e22f68839f332d829754c4c594a3ce9dd0/src/main/java/com/datadoghq/sketch/ddsketch/mapping/CubicallyInterpolatedMapping.java#L90

Reply:
I'll try to migrate `inverseLog` here.